save_flags.c

```diff
@@ -1,93 +1,72 @@
-#include <stdarg.h>
-#include <strings.h>
-#include <stdio.h>
-#define ARGS_BUF "cspdiouxXfeg#0-+ llhhlhLbrk'*$"
-#define ARGS_SIZE 32
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   save_flags.c                                       :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: fokrober <marvin@42.fr>                    +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2019/10/01 22:23:12 by fokrober          #+#    #+#             */
+/*   Updated: 2019/10/03 21:51:08 by fokrober         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
 
-int		ft_strncmp(char *s1, char *s2, int n);
-int		first_char_nbr(char *s);
+#include "ft_printf.h"
 
-void	save_flag(char *flags, char *format, int argw)
+int		save_flag(int *flags, char *format)
 {
-	char	*s;
 	int		pos;
+	int		fw;
 
-	s = ARGS_BUF;
-	pos = 0;
-	while (*s)
+	fw = first_char_nbr(format);
+	if ((pos = find_flag(FLAGS_BUF, format, fw)) == -1)
 	{
-		if (ft_strncmp(s, format, argw) == argw)
-		{
-			if (s[0] != s[argw])
-			{
-				flags[pos] = 1;
-				break ;
-			}
-			else
-				s += argw;
-		}
-		s += first_char_nbr(s);
-		pos++;
+		if (find_flag(FLAGS_BUF, format, 1) != -1)
+			return (fw);
+		return (0);
 	}
+	SET_FLAG_ON(*flags, pos);
+	return (fw);
 }
 
-int		ft_strnstr(char *s1, char *s2, int n)
+int		find_flag(char *flag_lst, char *format, int flagw)
 {
-	int		rep;
+	int		pos;
+	int		flagw_orig;
 
-	rep = n;
-	while (n)
+	pos = 0;
+	while (*flag_lst)
 	{
-		if ((*s1 && *s2))
-		{
-			if (*s1 != *s2)
-				return (0);
-		}
-		else
-			return (*s1 == *s2);
-		(void)(s1++ && s2++ && n--);
+		flagw_orig = first_char_nbr(flag_lst);
+		if (flagw == flagw_orig && ft_strncmp(flag_lst, format, flagw) == 0)
```

Comment:
doing that will not work coz: consider the case where the given flag is **h** and the string containing the flags looks like the one mentioned below.

```C
# define FLAGS_BUF "cspdiouxXfeg#0-+ llhhlhLbrk'*$"
```
when comparing each character once, we'll end up activating **hh** instead of **h**